### ISSUE TITLE
Documentation on library interface?

### ISSUE THREAD

**Issue by DoumanAsh** [ku-nlp/jumanpp#61]
Subj, would be nice to have it in form of doxygen or maybe short intro in README.

**Comment by eiennohito** [comment]
YES! I need to do this! Definitely! 

But right now I invite you to look into the main launcher https://github.com/ku-nlp/jumanpp/blob/master/src/jumandic/main/jumanpp.cc

**Comment by DoumanAsh** [comment]
Yes that's good starting point, though currently as I'm on windows I cannot use jumanpp.
Cuz #31

Though it gave me idea to write some cross-platform mmap for C++

**Comment by eiennohito** [comment]
Yes, if you can work on that it should be wonderful. 
The C++ itself code should be compilable with MSVC 2017 (I haven't tried), or definitely by Windows build of clang/mingw64.

There are 3 pieces of unix api I use, `mmap` (a wrapper is designed to be easily implemented for Windows though), `posix_align` (in util/memory, on Linux I use it with 2MB alignment and `madvice` to force transparent huge pages) and `unlink` in tests.

**Comment by eiennohito** [comment]
The other thing is, on Windows, filesystem access should use W versions of API (and convert paths from/to utf-8).

**Comment by eiennohito** [comment]
@DoumanAsh @kou what do you think.

**Comment by DoumanAsh** [comment]
@eiennohito Should it be then just API in general? After all there will be also C++ API? 🤔

**Comment by eiennohito** [comment]
C++ API won't have stability guarantees and won't support dynamic linking (I will try not to break it much though).

**Comment by DoumanAsh** [comment]
Ah that's understandable of course.

I guess the first thing we'd need to try to take a look at existing public API and see how it is to be used.
I understand goal of mimicking mecab API, but i think it wouldn't harm to do own thing if it can be better API.

I've somewhat busy lately but I'll try to take a look at existing API later on and give some inputs/suggestion on it.

**Comment by kou** [comment]
> C++ API won't support dynamic linking/systemwide installation as a library (basically, the current status-quo).

I'm a bit confused.
Does it mean that C API will support dynamic link but C++ API doesn't support dynamic link? If it's true, does C API use C++ API by static link?

**Comment by eiennohito** [comment]
Yes, C API will support consuming Juman++ as a dynamic object. I plan to build it with a bit different set of compile flags.

In theory it is possible to build even the current version Juman++ with `-fPIC` and link everything as dynamic libraries, and things will probably even work, but there could be a lot of not-very-fun problems with C++, dynamic linking, initialization order of globals and other runtime stuff.
E.g. exposing C++ classes over dll boundaries on Windows can lead to serious problems. 

The second problem is symbol visibility. We don't want to export our version of Eigen to outside world, because that would confuse other software using their own version of Eigen. On Unixes we should use `-fvisibility=hidden` for that and explicitly define exported symbols.

C API library will cleanly solve these problems by providing a single shared object with only C API marked as visible and all C++ state contained inside that object.

**Comment by kou** [comment]
Thanks. I understand what you explain.

**Comment by kou** [comment]
I'll try creating a C API and send feedback to Juman++ in a few months.

**Comment by eiennohito** [comment]
I maybe start doing that earlier, but I'll post here when I will begin doing it.

**Comment by eiennohito** [comment]
Nothing about C API yet, but I am writing a small tutorial on "How to use Juman++ to create your own morphological analyzers".

**Comment by kou** [comment]
Sorry. I couldn't work on this this year...
